src/Pact/Native/Db.hs

```diff
@@ -220,6 +237,30 @@ read' g0 i as@(table@TTable {}:TLitString key:rest) = do
 
 read' _ i as = argsError i as
 
+
+foldDB' :: GasNativeFun e
+foldDB' g0 i [tbl, TApp qry _, TApp consumer _] = do
+  table <- reduce tbl >>= \case
+    t@TTable{} -> return t
+    t -> evalError' i $ "Expected table as first argument to foldDB, got: " <> pretty t
+  (g1, ks) <- getKeys g0 table
+  (g2, xs) <- foldlM (fdb table) (g1, []) ks
+  pure (g2, TList (V.fromList (reverse xs)) TyAny def)
+  where
+  asBool (TLiteral (LBool satisfies) _) = return satisfies
+  asBool t = evalError' i $ "Unexpected return value from fold-db query condition " <> pretty t
+  getKeys g table = gasPostReads i g (map toTerm . sort) $ do
+    guardTable i table GtKeys
+    keys (_faInfo i) (userTable table)
+  fdb table (!g, acc) key = do
+    (!g1, row) <- read' g i [table, key]
+    cond <- asBool =<< apply qry [key, row]
+    if cond then do
+      r' <- apply consumer [key, row]
```

Comment:
Should there be a gas cost for applying `consumer` to the fetched row? That's still work we have to do?